Topology: Add simple topologies (no KWD) for CML RT1011 RT5682

singalsu opened the pull request:
This patch is based on sof-cml-rt1011-rt5682.m4 but omits the
hard coded keyword pipeline and resplaces it with a generic
16 kHz capture pipeline. The use of intel-generic-dmic.m4 to
set up the capture pipeline simplifies the topology a lot. This
change makes the topology for CML I2S quite similar to sof-hda-generic
to add audio processing.

All endpoint processing can be defined from top level CMakeLists.txt
with macros HSMICPROC, HSEARPROC, SPKPROC, DMICPROC, DMIC16KPROC.

A sample topology with equalizers in headset and speaker paths is
added.

Signed-off-by: Seppo Ingalsuo <seppo.ingalsuo@linux.intel.com>

singalsu commented:
@plbossart This changes PCM IDs of e.g. sof-cml-rt700 and sof-icl-rt700 topologies. I wonder if the previous IDs were a mistake? If no mistake, I need to add definitions for the old PCM IDs to those topologies.

plbossart commented on tools/topology/platform/intel/intel-generic-dmic.m4:
```diff
@@ -5,9 +5,9 @@
+ifdef(`DMIC_48k_PCM_NAME',`',
 `define(DMIC_48k_PCM_NAME, `DMIC')')
-ifdef(`DMIC_16K_PCM_NAME',`',
+ifdef(`DMIC_16k_PCM_NAME',`',
```
yes I fixed both in https://github.com/thesofproject/sof/pull/3840

plbossart commented on tools/topology/platform/intel/intel-generic-dmic.m4:
```diff
@@ -110,8 +114,8 @@ DAI_ADD(sof/pipe-dai-capture.m4,
-PCM_CAPTURE_ADD(DMIC_48k_PCM_NAME, DMIC_DAI_LINK_48k_ID, concat(`PIPELINE_PCM_', DMIC_PIPELINE_48k_ID))
-PCM_CAPTURE_ADD(DMIC_16k_PCM_NAME, DMIC_DAI_LINK_16k_ID, concat(`PIPELINE_PCM_', DMIC_PIPELINE_16k_ID))
+PCM_CAPTURE_ADD(DMIC_48k_PCM_NAME, DMIC_PCM_48k_ID, concat(`PIPELINE_PCM_', DMIC_PIPELINE_48k_ID))
+PCM_CAPTURE_ADD(DMIC_16k_PCM_NAME, DMIC_PCM_16k_ID, concat(`PIPELINE_PCM_', DMIC_PIPELINE_16k_ID))
```
fixed in https://github.com/thesofproject/sof/pull/3840

plbossart commented on tools/topology/platform/intel/intel-generic-dmic.m4:
```diff
@@ -79,7 +83,7 @@ ifdef(`DMIC16KPROC_FILTER1', `define(PIPELINE_FILTER1, DMIC16KPROC_FILTER1)', `u
 
 PIPELINE_PCM_ADD(sof/pipe-DMIC16KPROC-capture-16khz.m4,
-	DMIC_PIPELINE_16k_ID, DMIC_DAI_LINK_16k_ID, DMIC16K_PCM_CHANNELS, s32le,
+	DMIC_PIPELINE_16k_ID, DMIC_PCM_16k_ID, DMIC16K_PCM_CHANNELS, s32le,
```
this one looks needed, yes.

plbossart commented on tools/topology/platform/intel/intel-generic-dmic.m4:
```diff
@@ -66,7 +70,7 @@ ifdef(`DMICPROC_FILTER1', `define(PIPELINE_FILTER1, DMICPROC_FILTER1)', `undefin
 
 PIPELINE_PCM_ADD(sof/pipe-DMICPROC-capture.m4,
-	DMIC_PIPELINE_48k_ID, DMIC_DAI_LINK_48k_ID, DMIC_PCM_CHANNELS, s32le,
+	DMIC_PIPELINE_48k_ID, DMIC_PCM_48k_ID, DMIC_PCM_CHANNELS, s32le,
```
this one looks needed, yes.

plbossart commented on tools/topology/platform/intel/intel-generic-dmic.m4:
```diff
@@ -22,6 +22,10 @@ ifdef(`DMIC_DAI_LINK_48k_NAME',`',define(DMIC_DAI_LINK_48k_NAME, `dmic01'))
 
+# Need to define if not set
+ifdef(`DMIC_PCM_48k_ID',`', `define(DMIC_PCM_48k_ID, DMIC_DAI_LINK_48k_ID)')
+ifdef(`DMIC_PCM_16k_ID',`', `define(DMIC_PCM_16k_ID, DMIC_DAI_LINK_16k_ID)')
```
this is done differently in intel-generic-dmic-kwd.m4

````
ifdef(`DMIC_PCM_48k_ID',`',`errprint(note: Need to define dmic48k pcm id for intel-generic-dmic-kwd
)')
````

I am not sure it makes sense to set a default here, in general device numbers and dailinks are really different items.

@ranj063 can you chime in?

plbossart reviewed (commented):
Good timing Seppo, I was also doing some changes in https://github.com/thesofproject/sof/pull/3840

We should probably combine the patches after we get feedback from @ranj063 

I am not sure about the second patch though. Is this a test topology? What is the intent here? The kernel will not use it as is.

plbossart commented:
> @plbossart This changes PCM IDs of e.g. sof-cml-rt700 and sof-icl-rt700 topologies. I wonder if the previous IDs were a mistake? If no mistake, I need to add definitions for the old PCM IDs to those topologies.

@singalsu the intent is to have the following mapping.
DMIC: 10
DMIC16kHz: 11

It was broken on cml-rt700 and icl-rt700, we both found the same issue.

singalsu commented:
> I am not sure about the second patch though. Is this a test topology? What is the intent here? The kernel will not use it as is.

Our team has several of these devices in development use but it's rather difficult to customize the default topology due to the m4 files structure to test new algorithms etc. Kernel won't load this topology as such but I just copy the customized topology over the default topology. The -nokwd topology variant is much like sof-hda-generic-2ch topology, the microphone gain is set to usable level with +20 dB gain. The -nokwd-eq topology shows how various endpoints are complemented with (default flat) equalizers. I will add a topology version with mic beamformer. I couldn't do it to my current TDFB PR without this repair.

singalsu commented:
@plbossart Thanks for including my fixes to #3840 ! I'll drop the first part from this PR after your is merged.

plbossart commented:
> > I am not sure about the second patch though. Is this a test topology? What is the intent here? The kernel will not use it as is.
> 
> Our team has several of these devices in development use but it's rather difficult to customize the default topology due to the m4 files structure to test new algorithms etc. Kernel won't load this topology as such but I just copy the customized topology over the default topology. The -nokwd topology variant is much like sof-hda-generic-2ch topology, the microphone gain is set to usable level with +20 dB gain. The -nokwd-eq topology shows how various endpoints are complemented with (default flat) equalizers. I will add a topology version with mic beamformer. I couldn't do it to my current TDFB PR without this repair.

I think we need to move all these 'enhanced' topology either to a test directory or an 'experimental' one ('staging' might work as well). There's just too many versions (with eq, asrc, etc) that should not be part of the releases seen by users.

singalsu commented:
@plbossart The PR is now updated. This could be in other directory but where and how with minimal m4 and cmake rework? This version more suitable for running Ubuntu in the device (capture sensitivity is increased by 20 dB) and enables easy testing of new algorithms with new generated topologies due to cleaner structure. 

The released topology sof-cml-rt1011-rt5682.tplg could be cleaned up by using intel-generic-dmic-kwd.m4 but I'm hesitating to do that since it's a critical topology for ChromeOS.

lgirdwood reviewed (approved):
@singalsu do we still need to align this with @plbossart PR ? @ranj063 any comments ?

singalsu commented:
@lgirdwood When the topologies clean-up is done this topology should be moved to the DSP oriented topologies location. I think the EQ version with more EQs, after merged, should replace the topology sof-cml-eq-rt1011-rt5682.tplg that's currently used in CI. So there should be a CI test topologies location also. I'm OK to wait and change this patch after we have decided the new locations.

lgirdwood commented:
> @lgirdwood When the topologies clean-up is done this topology should be moved to the DSP oriented topologies location. I think the EQ version with more EQs, after merged, should replace the topology sof-cml-eq-rt1011-rt5682.tplg that's currently used in CI. So there should be a CI test topologies location also. I'm OK to wait and change this patch after we have decided the new locations.

@plbossart do you have a location change PR in progress now ? I'm looking to avoid conflicts for you in a PR that will churn, or can we merge now and move when you are ready ?

plbossart commented:
> > @lgirdwood When the topologies clean-up is done this topology should be moved to the DSP oriented topologies location. I think the EQ version with more EQs, after merged, should replace the topology sof-cml-eq-rt1011-rt5682.tplg that's currently used in CI. So there should be a CI test topologies location also. I'm OK to wait and change this patch after we have decided the new locations.
> 
> @plbossart do you have a location change PR in progress now ? I'm looking to avoid conflicts for you in a PR that will churn, or can we merge now and move when you are ready ?

@lgirdwood @singalsu no, nothing in progress for now on my side. I'd rather let Seppo handle his DSP processing examples.

lgirdwood commented:
SOFCI TEST

lgirdwood commented:
Rerun CI again as a CML device did not boot.

xiulipan reviewed (approved):
A simple question about the usage of this topology. we have a folder `tools/topology/development` for the debug and developing tplg.

xiulipan commented on tools/topology/CMakeLists.txt:
```diff
@@ -132,6 +132,8 @@ set(TPLGS
 	"sof-cml-demux-rt5682-max98357a\;sof-cml-demux-rt5682-max98357a\;-DPLATFORM=cml"
 	"sof-cml-rt1011-rt5682\;sof-cml-rt1011-rt5682\;-DPLATFORM=cml\;-DPPROC=volume"
 	"sof-cml-rt1011-rt5682\;sof-cml-eq-rt1011-rt5682\;-DPLATFORM=cml\;-DPPROC=eq-iir-eq-fir-volume"
+	"sof-cml-rt1011-rt5682-nokwd\;sof-cml-rt1011-rt5682-nokwd\;-DCHANNELS=2\;-DDMICPROC_FILTER1=eq_iir_coef_highpass_40hz_20db_48khz.m4\;-DDMIC16KPROC_FILTER1=eq_iir_coef_highpass_40hz_20db_16khz.m4"
```
Do we have machine driver use the tplg by its name `sof-cml-rt1011-rt5682-nokwd.tplg`

singalsu replied:
No, this is not loaded by machine driver, the plain sof-cml-rt1011-rt5682.tplg is used so this needs to be copied over  it. This could go to development, I'll check how it works for me.

singalsu replied:
@xiulipan I just updated the PR with this topology moved to development. There was no changes needed to topology so this works well.

xiulipan replied:
@singalsu Thanks for the update. I think we can do some clean up to move some experiment audio processing tplg to development folder.

lgirdwood commented:
@singalsu is this aligned with your new topology directory structure ?

singalsu commented:
> @singalsu is this aligned with your new topology directory structure ?

Yes it is. This patch adds the topology to development.

singalsu commented:
I just rebased to fix the merge conflict. I renamed the new topology variant from "...nokwd-eq..." to "...eq..." for simpler topology name. There will be later e.g. "...aec16k..." and "...aec48k..." for this platform similarly and I didn't feel like repeating "nokwd" every time is useful.

lgirdwood commented:
@marc-hb @xiulipan - can you guys comment on the CI checktree error - it's not obvious to me.

marc-hb commented:
The installer checktree error in https://github.com/thesofproject/sof/pull/3841/checks?check_run_id=2021208085 has nothing to do with the installer, it's a docker timeout: `net/http: TLS handshake timeout`

One hand it can be ignored because merely adding some topologies can not possibly make this test fail.

On the other hand we should probably get worried about docker limits in general because Travis has recently been hitting docker rate limits frequently. @xiulipan do we use docker tokens anywhere?

lgirdwood commented:
@xiulipan @marc-hb I'm assuming we only get a certain amount of Docker cycles per week/month ? Is there a link to any SLA ?

marc-hb commented:
There is some info at https://github.community/t/did-dockerhub-rate-limit-affect-github-action/128158
